lib/process/child-pool.js

```diff
@@ -3,73 +3,57 @@
 var fork = require('child_process').fork;
 var path = require('path');
 var Promise = require('bluebird');
+var _ = require('lodash');
 
 module.exports = function ChildPool() {
   if(!(this instanceof ChildPool)){
     return new ChildPool();
   }
 
-  var retained = {};
-  var free = [];
+  this.retained = {};
+  this.free = [];
 
-  this.retain = function(processFile){
-    return new Promise(function(resolve, reject) {
+  this.retain = Promise.method(function(processFile){
+    var child = this.free.pop();
 
-      var child = free.pop();
-      if(child){
-        return resolve(child);
-      }
+    if (child) return child;
 
-      try{
-        var child = fork(path.join(__dirname, './master.js'));
+    child = fork(path.join(__dirname, './master.js'));
 
-        retained[child.id] = child;
+    this.retained[child.pid] = child;
 
-        child.on('exit', function(code, signal){
-          console.error('Child process exited', child.pid, code, signal);
+    child.on('exit', this.remove.bind(this, child));
 
-          // Remove exited child
-          deleteChild(child);
+    var send = function(msg) {
+      return new Promise(function(resolve){
+        child.send(msg, resolve);
+      });
+    };
 
-          function deleteChild(child){
-            delete retained[child.pid];
-            var childIndex = free.indexOf(child);
-            if (childIndex > -1) {
-              free.splice(childIndex, 1);
-            }
-          }
-        });
+    return send({ cmd: 'init', value: processFile }).return(child);
+  });
 
-        child.send({
-          cmd: 'init',
-          value: processFile
-        }, function() {
-          resolve(child);
-        });
-      } catch(err){
-        reject(err);
-      }
-    });
+  this.release = function(child){
+    delete this.retained[child.pid];
+    this.free.push(child);
   };
 
-  this.release = function(child){
-    delete retained[child.pid];
-    free.push(child);
+  this.remove = function(child){
+    delete this.retained[child.pid];
+    var childIndex = this.free.indexOf(child);
+    if (childIndex > -1) this.free.splice(childIndex, 1);
+  };
+
+  this.kill = function(child){
+    child.kill();
```

Comment:
I think it would be useful to pass the kill signal as well, for future use: 
https://nodejs.org/api/child_process.html#child_process_subprocess_kill_signal